jax/_src/lax/other.py

```diff
@@ -114,3 +114,115 @@ def conv_general_dilated_patches(
       preferred_element_type=preferred_element_type
   )
   return out
+
+
+def conv_general_dilated_local(
+    lhs: jnp.ndarray,
+    rhs: jnp.ndarray,
+    window_strides: Sequence[int],
+    padding: Union[str, Sequence[Tuple[int, int]]],
+    filter_shape: Sequence[int],
+    lhs_dilation: Optional[Sequence[int]] = None,
+    rhs_dilation: Optional[Sequence[int]] = None,
+    dimension_numbers: Optional[convolution.ConvGeneralDilatedDimensionNumbers] = None,
+    precision: Optional[lax.PrecisionLike] = None
+) -> jnp.ndarray:
+  """General n-dimensional unshared convolution operator with optional dilation.
+
+  Also known as locally connected layer, the operation is equivalent to
+  convolution with a separate (unshared) `rhs` kernel used at each output
+  spatial location. Docstring below adapted from `jax.lax.conv_general_dilated`.
+
+  See Also:
+    https://www.tensorflow.org/xla/operation_semantics#conv_convolution
+
+  Args:
+    lhs: a rank `n+2` dimensional input array.
+    rhs: a rank `n+2` dimensional array of kernel weights. Unlike in regular
+      CNNs, its spatial coordinates (`H`, `W`, ...) correspond to output spatial
+      locations, while input spatial locations are fused with the input channel
+      locations in the single `I` dimension, in the order of
+      `"C" + ''.join(c for c in rhs_spec if c not in 'OI')`, where
+      `rhs_spec = dimension_numbers[1]`. For example, if `rhs_spec == "WHIO",
+      the unfolded kernel shape is
+      `"[output W][output H]{I[receptive window W][receptive window H]}O"`.
```

Comment:
Why not make the argument `2*n+2` dimensional, e.g., (w, h, x, y, i, o) ?

Reply:
I agree the shape is not intuitive, but the reasoning here was to avoid redundant reshapes. Namely, `conv_general_dilated_patches` returns flattened patches (due to using a conv-based implementation - https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.conv_general_dilated_patches.html), hence I wanted the local convolution to have weights that can be multiplied with patches without reshapes.

If we make `rhs` higher-dimensional, then each forward/backward pass would have a reshape of `rhs` before applying. One alternative is to have `conv_general_dilated_patches` itself return unflattened patches, but then the reshapes will be happening inside `conv_general_dilated_patches`.

I know that some reshapes are easily optimized away, but some aren't and can introduce a performance hit (see e.g. b/168747617), so I thought it would be safer to keep it without reshapes. Lmk if this makes sense, or do you think it won't impact performance here?

Reply:
Personally I think both functions would be clearer if they took/returned higher dimensional arrays, but I see your point.